### DISCUSSION TITLE
Получение IP-адресов для связки Keenetic-Wireguard

### DISCUSSION THREAD

**Discussion by GedLouZe** [Ground-Zerro/DomainMapper#58] [Q&A] [answered]
Доброго времени суток!
_Сразу хочу извиниться за неточности и возможные непонятки в моем разъяснении проблемы_
_Я понимаю, что лезу в область, в которую по сути лезть не должен, однако времена нынче такие и хочется все сделать по уму_

Сначала о спеках, потом к вопросам

**Локация**: Санкт-Петербург, Ленинградская область
**Провайдер**: Ростелеком | Енева (резервный)
**Оборудование**: 
1. Huawei HG8120H для оптики Ростелекома (режим моста)
2. Keenetic Speedster KN-3012

**Подключение**:
В Keenetic Speedster KN-3012 приходит два интернета, Ростелеком из моста Huawei HG8120H, Енева напрямую (резервный)
На выходе два ПК по Ethernet и куча мобилок по Wi-Fi

**Настройка Keenetic**:
IP-адрес: статичный от провайдера (Ростелеком)
DNS-провайдера отключены
DNS добавлены через DoT

**Wireguard**: VPS-сервер в Латвии

**Проблематика**:
Необходимо пробросить различные сервисы через Wireguard, что бы весь остальной трафик шел напрямую через провайдеров

**Предпринятые попытки решения**:
1. Собрал IP-адреса для Wireguard (по схеме все сервисы, DNS Cloudflare, сокращение до /16), внес их в конфиг в строку AllowedIPs. 
**Итог - интернет полностью отваливается**

2. Собрал чистые IP-адреса для Keenetic (по схеме все сервисы, все DNS, сокращение до /16), конфигурировал их в .bat файл для внесению в таблицу маршрутизации. 
**Итог**:
- YouTube работает на ПК, долго прогружает страницы с видосом, но так или иначе тащит, Shorts не грузит совсем
- YouTube на мобилках работает хуже чем на ПК, Shorts не грузит совсем
- Soundcloud работает замечательно
- Discord работает, но пинг достигает 100 мс
- только что понял, что весь трафик гонится через Wireguard

**Вопрос**:
Как правильно собрать IP-адреса и как лучше их пробрасывать (через роутер или чере Wireguard)

Буду благодарен за любую помощь, да и просто за желание помочь.

**Comment by Ground-Zerro** [comment, marked as answer]
Сразу скажу, что мое мнение не является авторитетным и лишь отражает мое личное понимание и личный опыт по высказанным вопросам.

**Начнем с самого начала.**
Speedster KN-3012 роутер без USB, соответственно среда Entware не доступна. Значит варианты с "маршрутизацией на основе доменных имен" на самом роутере отпадают.

**Настройка роутера**
Чтобы избежать влияние проведенных ранее манипуляций.
- Удалить все добавленные ранее статически маршруты: `Маршрутизация -> Удалить все маршруты`.
- Удалить все добавленные ранее WG туннели.
- Удалить все ранее созданные политики доступ: `Приоритеты подключений`, оставить только `политику по умолчанию`.
- На VPS создать нового пира (пользователя) Wireguard.
- Открыть `conf` файл созданного пира (пользователя) и удалить строчку с DNS (больше ничего не менять)
Пример, было:
```
[Interface]
PrivateKey = KByCcJgCkQJkYEj4F6b/7ZV09Ul6sRBVSc/i5BVUqUUw=
Address = 10.231.32.2/24
DNS = 9.9.9.9, 149.112.112.112

[Peer]
PublicKey = uK6454xs/FuU4jrbKjfRte5rtah0ELkjAaC5YTvbwE=
PresharedKey = qY2IyEYZSoCgJhGbvCfgtR1tuDc27QDVDOswqUr9M=
Endpoint = 121.45.41.54:3456
AllowedIPs = 0.0.0.0/0, ::0/0
```
стало:
```
[Interface]
PrivateKey = KByCcJgCkQJkYEj4F6b/7ZV09Ul6sRBVSc/i5BVUqUUw=
Address = 10.231.32.2/24

[Peer]
PublicKey = uK6454xs/FuU4jrbKjfRte5rtah0ELkjAaC5YTvbwE=
PresharedKey = qY2IyEYZSoCgJhGbvCfgtR1tuDc27QDVDOswqUr9M=
Endpoint = 121.45.41.54:3456
AllowedIPs = 0.0.0.0/0, ::0/0
```
- Сохранить файл и добавить WG на роутер через кнопку `загрузить из файла`.
    - ничего в настройках созданного в роутере подключения не менять, галочки никакие не ставить.
    - включить WG чтобы проверить есть ли подключение и рукопожатие.
    ![123](https://github.com/user-attachments/assets/e9c43d4d-18c2-4269-bae6-1fbe9cf6618c)
    - если рукопожатия нет, значит нужно сначала решить эту проблему. Либо VPS настроен неверно, либо софт на нем, либо еще что-то.

**Дальше сценарии могут быть разными.**
Самый простой это статические маршруты.... (будь они трижды не ладны!!!)
- Используя [DomainMapper](https://github.com/Ground-Zerro/DomainMapper/tree/main/Windows) создать файл маршрутов.
    - запускаем, выбираем только нужные сервисы, выбираем только !!!СИСТЕМНЫЙ DNS!!!, `cloudeflare` пусть останется.
    - агреация в подсети `mix`
        - главное чтобы получилось не больше 1024 маршрута, иначе попробовать /24, если тоже будет больше, значит переделать под /16
    - формат сохранения `win`, шлюзом указать `0.0.0.0`
    - **ВАЖНО запускать DM нужно на машине которая находится за твоим роутером. НЕ на VPS, не просить соседа или ще кого-то, сделать на своей машине за роутером и никак по другому**.
    - добавить сформированный файл маршрутов в роутер: `маршрутизация -> загрузить из файла`.
    ![234](https://github.com/user-attachments/assets/91b8bac5-f57f-4b30-bbd2-4ac842b6d270)

Видео с примером [ЖМИ](https://github.com/Ground-Zerro/DomainMapper/issues/47#issuecomment-2428952266).
Это только пример! Сервисы выбрать под себя, DNS - системный и т.д.

***Все***
WG настроен, подключен и работает.
Маршруты прописаны.
Трафик до сервисов идет через VPS.

Главный минус способа то, что IP адреса у сервисов меняются и маршруты нужно постоянно обновлять. То что откроется на одном ПК может уже не открыться на втором (ПК это те, что находятся за роутером).


**Более сложный но чуть более правильный вариант в случае с KN-3012**
В свое время для выхода из этой ситуации написал [DNS-RouteSync-Navigator](https://github.com/Ground-Zerro/DNS-RouteSync-Navigator) который относительно успешно решал задачу роутинга на основе DNS имен. Учитывая что у вас арендован VPS такой вариант вполне может подойти. Инструкция и файлы доступны по ссылке. Рекомендую взять версию с DoH шифрованием.

**Reply by GedLouZe** [reply]
Круто, оч подробно и даже мне "лбу" понятно, пока что :)

Спасибо мил человек, низкий тебе поклон!

**Reply by GedLouZe** [reply]
@Ground-Zerro попробовал первый способ
сократить получилось только до /16, в итоге получил ~700 строк
работоспособность осталась на прежнем уровне
провел опыт, работоспобность чистого подключение через WG в разы стабильнее подключения по маршрутам

Начинаю вчитываться во второй способ, как попробую сообщу

**Reply by Ground-Zerro** [reply]
> работоспособность осталась на прежнем уровне

Подозрение, что где-то не так прописаны DNS.
Некоторые браузеры, например Яндекс, использует свои "Безопасные DNS" и они включены в нем по умолчанию.
В этом случае домены или их часть в браузере резолвятся не на те IP, что прописаны в маршрутах статикой.
![345](https://github.com/user-attachments/assets/8237eb47-4547-4e5c-b8cf-bf83a8d13502)



В роутере как отключены DNS провайдера?
- галочкой `Игнорировать DNSv4 провайдера`
![111](https://github.com/user-attachments/assets/94579fd3-1446-49ff-8635-6fc8ce71489f)
Или как-то еще?



Еще вариант:
На ПК в настройках сети указан IP DNS сервера отличный от выдаваемого роутером, обычно роутер дает `192.168.1.1`.


Еще вариант:
При манипуляциях добавил маршруты на ПК.
Как првоерить:
- Откройте командную строку с правами администратора.
- Введите команду:
`route print`
Должно быть пусто:
![route](https://github.com/user-attachments/assets/3888769a-0f73-4c6f-9096-2d0a05ebaf34)


Еще вариант:
Включен IPv6.

**Reply by GedLouZe** [reply]
Браузер использую Waterfox 6.5.1

с DNS ситуация такая

![изображение](https://github.com/user-attachments/assets/215c731e-3c58-4998-8c87-e3750180557e)

![изображение](https://github.com/user-attachments/assets/e6689156-be05-49f5-8099-938ff012b64b)

Дополнено:

При сборе маршрутов выбирал системный
![изображение](https://github.com/user-attachments/assets/19ac63f7-df8b-4428-b730-694920428c9a)

Вывод `route print`
![изображение](https://github.com/user-attachments/assets/4521ab07-e891-448b-a4ff-988b379ea9ed)

IPv6 в настройках адаптера ПК был включен, галочку снял

**Comment by Ground-Zerro** [comment]
IPv6 от провайдера???

Так не получится.
IPv6 имеет приоритет над IPv4.
Маршруты **в роутере** прописаны только для IPv4.
Вот и тупит.
IPv6 в роутере - отключить.

Примерно как вижу:
Когда прямо указываешь в роутере политикой пускать устройство только через WG - IPv6 отключается и все хорошо.
Мало какие хостеры дают VPS с IPv6 выше /64 которые можно разбить чтобы маршрутизировать трафик с хостов, мало какой софт поднимает WG туннели с IPv6 без дополнительной настройки.

**Reply by GedLouZe** [reply]
в настройках роутера IPv6 отключил

![изображение](https://github.com/user-attachments/assets/1d070d5a-c689-48de-be4b-b71b5cc2ab79)

По итогу осталось все так же, задержка в Discord снизилась

**Reply by Ground-Zerro** [reply]
Это было основной проблемой думаю. Сделай маршруты по новой, в роутере удали те что добавил и залей новые.

**Reply by Ground-Zerro** [reply]
По Дискорду - DM пускает через VPS все голосовые сервера в т.ч. те что в РФ.

По задержкам, выполни тест на VPS:
```
wget -qO- https://raw.githubusercontent.com/Ground-Zerro/My_trivia/refs/heads/main/bench.sh | bash
```
И все станет ясно.



Заметка для меня: Добавить в бэнчь РФ ST сервера....

**Reply by GedLouZe** [reply]
Добро, как все прокину еще раз отпишусь

По тесту такой вывод
![изображение](https://github.com/user-attachments/assets/a5844054-0b08-42e4-b4b0-91816646d9e4)

**Reply by Ground-Zerro** [reply]
Ну в целом то не плохо.

Справа задержки от VPS к сервакам СпидТеста в разных странах, добавь к ним задержку от себя к твоему VPS и будет минимально возможная (образно - в вакууме) задержка до серверов Дискорда.

Кстати IPv6 на VPS нет.
И сменить у сервера TCP CC на `bbr` вместо `cubic` будет не лишним. [Подробней здесь.](https://www.k7d.ru/linux/vklyuchenie-tcp-bbr-uskoryaem-set-v-ubuntu-linux/)

**Reply by GedLouZe** [reply]
спасибо, сделал
![изображение](https://github.com/user-attachments/assets/2ebee90c-1240-48a0-b22c-e82f04f9766b)

ожидаю возможности пересобрать маршрутизацию, интернет занят супругой, не хочу обрушить на себя ее гнев :D

**Comment by Ground-Zerro** [comment]
Думаю помог чем смог, больше вряд-ли подскажу чего.
Единственно отмечу, что 700 маршрутов многовато для твоего кинетика, он вполне может нагружаться обрабатывая таблицу маршрутизации и тупить отсюда и задержки.

Удачи.

Если "вдруг", есть [ТГ группа](https://t.me/+iy38m3CTIGExODcy) для обратной связи, там общаться поудобней. Создал когда-то, но ссылок не публиковал, я пока там один )))

**Reply by GedLouZe** [reply]
И так, все пересосбрал, результат тот же, к сожалению, а так же начала повторятсья болячка, что 2ip видит меня на локации VPS :)

Тем не мене, остался еще один способ, над ним и буду корпеть на днях :)

Спасибо тебе огромное, за твою работу, помощь и отзывчивость! Надеюсь ты потратил на меня время не зря и у тебя появились идеи для будущий улучшений и начинаний!

За сим не прощаюсь, по второму способу как только так сразу отпишусь :)

**Comment by Ground-Zerro** [comment]
_Вчера в ТГ написал, продублирую здесь._

Думаю я понял в чем проблема. И как сразу не сообразил...

Судя по картинке в роутер напихана куча разных публичных DNS.

**=== Далее все очень условно в качестве примера, для понимания происходящего ===**

Роутер пытается распределить нагрузку между DNS серверами равномерно и каждый следующий DNS запрос отправляет к разным DNS серверам по порядку (либо использует их по своему алгоритму но не один и тот-же все время).

В чем соль: для условного `site.com` GoogleDNS вернёт `100.20.20.10`, Quad9DNS вернёт `30.44.33.77`, CloudflareDNS вернет `66.77.22.45` и т.д.

А маршрутов то к ним нет т.к. при резолве доменов DomainMapper получал IP-адреса от разных DNS.

Получается такая картинка: для `site.com` в DM вернулся IP от GoogleDNS, а когда `site.com` открывают с ПК ответ возвращается от CloudeflareDNS с совершенно другим IP-адресом. При повторном открытии `site.com` с ПК ответ может вернуться от GoogleDNS и тогда все **Ок**, а может и от Quad9DNS и тогда снова мимо.

Учитывая что при открытии в браузере одного только `youtube.com` уходит куча DNS запросов (картинки, комментарии, превью, видео, реклама и т.д. - на все могут быть свои отдельные обращения к DNS) вот и получается каша. Маршруты добавлены для одного IP, а DNS вернул другой, часть трафика пошла через туннель, часть напрямую, что-то работает на странице, а что-то нет, обновил страницу и все поменялось.

**Что сделать, по порядку:**
- убрать все DNS серверы из роутера оставив только один (или пару НО от какого-то одного сервиса с одинаковыми настройками безопасности, приватности и фильтрации), например только [Quad9DNS](https://en.wikipedia.org/wiki/Quad9) ну или другой.
- удалить из роутера добавленные ранее маршруты.
- Пересоздать маршруты а DM используя Системный DNS (будет использован тот один DNS сервер, который остался в роутере).
- залить новые маршруты в роутер.


> 2ip видит меня на локации VPS

Подсети у тебя сокращены до /16 маски, IP-адрес `2ip.ru` = 188.40.167.82 что входит в 188.40.0.0/16 подсеть или диапазон между 188.40.0.0 - 188.40.255.254, а это `65 тысяч 534` узла.
Поэтому и говорю не пихать в список ВСЕ подряд, а ограничиться только нужными сервисами.
Разграничить `RU` зону и все остальное таким образом - не получится! На двух даже соседних IP легко могут быть узел заблокированный РКН и какие-нибудь муниципальные услуги города Энска. Иногда они вообще соседствуют на одном IP. А люди шинкуют подсети до 16 маски, это ооочень много.

**Reply by GedLouZe** [reply]
ТГ увидел, но не раньше этого сообщения, спасибо! :)

Оставил только [Quad9DNS](https://en.wikipedia.org/wiki/Quad9)

![изображение](https://github.com/user-attachments/assets/2a0bfc94-1071-4f6c-bbc1-d975488b8073)

1. Сделал сервисов по минимуму, так или иначе в mix всеравно вылезает 1600+ адресов, однако в этот раз я подумал "была не была!" и зарядил в маршрутизацию ВСЕ адреса, банально поделив пополам между батниками (которые более 1024 строк не держат)
**В общем, маршруты я завел, Youtube стал работать бодрее, на мобилках совсем не работает, Discord отвалился** :D

2. Аналогично п. 1, только mix заменил на 24, получил примерно столько же адресов, так же поделил налопопам и залил в роутер
**В сравнении с п. 1, Youtube работает аналогично, Discord завелся!** :D

**Comment by Ground-Zerro** [comment]
Да, иногда это мера вынужденная, до /16 жать.
Дело в том, что роутер больше 1024 маршрутов не умеет, банально в таблицу не влазит. Админка говорит все Ок, но если поискать то их не будет в списке. По-моему на какой-то прошивке это правили, точно не скажу, а врать не буду.

Заработало?
Разобрались стало быть и победили ))
Вот и хорошо.

**Reply by GedLouZe** [reply]
Я проверил, посчитал всю таблицу маршрутизации, вышло столько, сколько закидывал (1600+), вопрос в том, а работают ли они :)

Получается в данной постановке вопроса мы исчерпали все возможные варианты :D 
Вышло не так идеально, как я себе это представлял, но "лучшее, враг хорошего" :)
Зато теперь, в случае чего, буду знать, что сделать что бы хоть как то работало :D
